### ISSUE TITLE
PMIx MPI support in Slurm

### ISSUE THREAD

**Issue by tpdownes** [GoogleCloudPlatform/hpc-toolkit#2274]
Hello!
 
I'm trying to deploy a cluster with pmix installed as a slurm plugin using the "image-builder" blueprint ( https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/develop/examples/image-builder.yaml ) and I can seem to get the image to build/deploy correctly.

The communication between the controller and login nodes keeps getting messed up, something similar to this https://groups.google.com/g/slurm-users/c/Q7FVkhx-bOsand this https://bugs.schedmd.com/show_bug.cgi?id=10066.
 
I'm guessing that I'm injecting my pmix installation stuff at the wrong moment (https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/develop/examples/image-builder.yaml#L42). Also, it requires re-configuring and installing slurm so I would want to run this at the very end of everything getting setup.

My current guess is that I would actually need to make an Ansible task and follow something similar to how the “install_ompi” tasks are setup through Ansible?

https://github.com/GoogleCloudPlatform/slurm-gcp/blob/849107c9f2e8044051877de579c7308e4e655139/packer/main.pkr.hcl#L31

Is this something that is on the roadmap or possibly something that you all have done before?

_Originally posted by @tomschultz-roivant in https://github.com/GoogleCloudPlatform/hpc-toolkit/discussions/2253_

**Comment by tpdownes** [comment]
To summarize #2253 reported by @tomschultz-roivant:

- the Slurm-GCP v5 images do not contain PMI support
- support can be added in the v5 modules by installing PMIx development libraries before compiling slurm following the [Slurm image building example](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/4dbcd38a16b2f23247e40105802f7d19b628e8b0/community/examples/hpc-build-slurm-image.yaml) (see below for a blueprint snippet)
- More actions will be required for v6 modules because [Slurm 23.11 no longer builds PMIx by default even if it is present](https://github.com/SchedMD/slurm/blob/19fb4b89d4d719641924067201c032b86ebeafe3/NEWS#L589)

```
blueprint_name: hpc-build-slurm-image

vars:
  project_id:  ## Set GCP Project ID Here ##
  deployment_name: build-slurm-1
  region: us-central1
  zone: us-central1-a

  image_build_machine_type: n2d-standard-16
  build_from_image_family: hpc-centos-7
  build_from_image_project: cloud-hpc-image-public
  build_from_git_ref: 5.10.4
  built_image_family: my-custom-slurm
  # built_instance_image:
  #   family: $(vars.built_image_family)
  #   project: $(vars.project_id)
  # instance_image_custom: true

deployment_groups:
- group: setup
  modules:
  - id: network
    source: modules/network/vpc

  - id: slurm-build-script
    source: modules/scripts/startup-script
    settings:
      # Do not create Ansible virtual env; Install system wide Ansible below.
      install_ansible: false
      runners:
      - type: shell
        destination: prep-for-slurm-build.sh
        content: |
          #!/bin/bash
          set -e -o pipefail
          # Slurm build on Rocky8 will upgrade to python38 as part of build
          # This is not compatible with ansible-local runner
          yum install -y python3
          yum install -y pmix
          # alternatives --set python3 /usr/bin/python3
          python3 -m pip install pip --upgrade
          python3 -m pip install ansible
          python3 -m pip install selinux
          export PATH=/usr/local/bin:$PATH
          ansible --version
          ansible-galaxy role install googlecloudplatform.google_cloud_ops_agents
      - type: data
        destination: /var/tmp/slurm_vars.json
        content: |
          {
            "reboot": false,
            "install_cuda": false,
            "nvidia_version": "latest",
            "install_ompi": true,
            "install_lustre": false,
            "install_gcsfuse": true
          }
      - type: shell
        destination: install_slurm.sh
        content: |
          #!/bin/bash
          set -e -o pipefail
          ansible-pull \
              -U https://github.com/GoogleCloudPlatform/slurm-gcp -C $(vars.build_from_git_ref) \
              -i localhost, --limit localhost --connection=local \
              -e @/var/tmp/slurm_vars.json \
              ansible/playbook.yml

- group: build-slurm
  modules:
  - id: slurm-custom-image
    source: modules/packer/custom-image
    kind: packer
    settings:
      machine_type: $(vars.image_build_machine_type)
      source_image_family: $(vars.build_from_image_family)
      source_image_project_id: [$(vars.build_from_image_project)]
      image_family: $(vars.built_image_family)
    use:
    - network
    - slurm-build-script
```

**Comment by tpdownes** [comment]
Responding to https://github.com/GoogleCloudPlatform/hpc-toolkit/discussions/2253#discussioncomment-8572391

I suspect your issue is that you are combining pieces of 2 different blueprints and missing a setting here or there. I will suggest a complete blueprint in the next response shortly after I test it

**Comment by tpdownes** [comment]
I believe the issue you're reporting in https://github.com/GoogleCloudPlatform/hpc-toolkit/discussions/2253#discussioncomment-8572391 may be that the default partition (debug) is not using the PMIx-enabled image. Your compute partition has the correct configuration. You can test the same commands on a running cluster with `srun -p compute ....` where the dots represent the same flags you used without specifying the partition.

I'm still testing end-to-end blueprint but this suggestion might resolve your immediate issue.

**Comment by tpdownes** [comment]
@tomschultz-roivant for some reason I'm having a bit of trouble re-building the image from the blueprint snippet, but I do think your issue can be resolved by ensuring the default "debug" partition uses the same image as the rest of the cluster using 

```
      instance_image_custom: true
      instance_image:
        family: my-custom-slurm # Here is where I reference the images from the other script after I built them and confirmed they were in the GCP compute image list
        project: <project_id>
```

and/or specifying the "compute" partition when you run commands. This is the partition you're more likely to use for application workloads anyhow.

**Comment by tomschultz-roivant** [comment]
Sorry for my slow response. I have been trying a few things and sadly, I haven't made any breakthroughs.

For the image build script, I also tried adding `yum install -y pmix pmix-devel` instead of manually downloading pmix and building it.  Installing pmix with yum didn't appear to be picked up automatically and didn't put anything un /usr/local which I believe is required for slurm to pick it up.


>If PMIx was installed to a standard location (i.e., with a prefix of /usr or /usr/local), then Slurm will find it and build the PMIx plugin by default. Otherwise, Slurm should be configured with the –with-pmix=path-to-where-pmix-was-installed option. The PMIx plugin will be built and installed under the Slurm location.

Ref: https://openpmix.github.io/support/how-to/slurm-support


Building pmix from source ensured that the files were in that location but slurm (23.02.7) again didn't appear to pick it up.

My next attempt will be to modify the ansible playbook to add the `--with-pmi=/usr/local/pmix` flag.

I have been testing by running `srun --mpi=list` on the login node and then `srun -p compute -N 1 -n 1 srun --mpi=list` to check on the compute nodes.  `pmix` doesn't show up in any case so far.


**Image Build Blueprint:**
```
blueprint_name: hpc-build-slurm-image

vars:
  project_id:  ## Set GCP Project ID Here ##
  deployment_name: build-slurm-1
  region: us-central1
  zone: us-central1-a

  image_build_machine_type: n2d-standard-16
  build_from_image_family: hpc-centos-7
  build_from_image_project: cloud-hpc-image-public
  build_from_git_ref: 5.10.4
  built_image_family: my-custom-slurm
  # built_instance_image:
  #   family: $(vars.built_image_family)
  #   project: $(vars.project_id)
  # instance_image_custom: true

deployment_groups:
- group: setup
  modules:
  - id: network
    source: modules/network/vpc

  - id: slurm-build-script
    source: modules/scripts/startup-script
    settings:
      # Do not create Ansible virtual env; Install system wide Ansible below.
      install_ansible: false
      runners:
      - type: shell
        destination: prep-for-slurm-build.sh
        content: |
          #!/bin/bash
          set -e -o pipefail
          # Slurm build on Rocky8 will upgrade to python38 as part of build
          # This is not compatible with ansible-local runner
          yum install -y python3 wget libevent-devel
          # alternatives --set python3 /usr/bin/python3
          python3 -m pip install pip --upgrade
          python3 -m pip install ansible
          python3 -m pip install selinux
          export PATH=/usr/local/bin:$PATH

          # Now manually install pmix in the location that slurm expects
          # Slurm defaultly looks for pmix in /usr/local/pmix
          mkdir -p /usr/local/src/pmix
          cd /usr/local/src/pmix

          # Download and install pmix
          wget https://github.com/openpmix/openpmix/releases/download/v4.2.4/pmix-4.2.4.tar.gz
          tar -xzvf pmix-4.2.4.tar.gz
          cd pmix-4.2.4

          # Slurm defaultly looks for pmix in /usr/local/pmix
          mkdir /usr/local/pmix

          ./configure --prefix=/usr/local/pmix |& tee config.out
          make -j 8 |& tee make.out
          make install |& tee install.out

          ansible --version
          ansible-galaxy role install googlecloudplatform.google_cloud_ops_agents
      - type: data
        destination: /var/tmp/slurm_vars.json
        content: |
          {
            "reboot": false,
            "install_cuda": false,
            "nvidia_version": "latest",
            "install_ompi": true,
            "install_lustre": false,
            "install_gcsfuse": true
          }
      - type: shell
        destination: install_slurm.sh
        content: |
          #!/bin/bash
          set -e -o pipefail
          ansible-pull \
              -U https://github.com/GoogleCloudPlatform/slurm-gcp -C $(vars.build_from_git_ref) \
              -i localhost, --limit localhost --connection=local \
              -e @/var/tmp/slurm_vars.json \
              ansible/playbook.yml

- group: build-slurm
  modules:
  - id: slurm-custom-image
    source: modules/packer/custom-image
    kind: packer
    settings:
      machine_type: $(vars.image_build_machine_type)
      source_image_family: $(vars.build_from_image_family)
      source_image_project_id: [$(vars.build_from_image_project)]
      image_family: $(vars.built_image_family)
    use:
    - network
    - slurm-build-script
 ```

**Cluster Blueprint:**
```
blueprint_name: psi-slurm-simple

vars:
  project_id:  ## Set GCP Project ID Here ##
  deployment_name: psi-slurm-simple-test
  region: us-central1
  zone: us-central1-a

# Documentation for each of the modules used below can be found at
# https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/main/modules/README.md

deployment_groups:
- group: primary
  modules:
  # Source is an embedded resource, denoted by "resources/*" without ./, ../, /
  # as a prefix. To refer to a local resource, prefix with ./, ../ or /
  # Example - ./resources/network/vpc
  - id: network1
    source: modules/network/vpc

  - id: homefs
    source: modules/file-system/filestore
    use: [network1]
    settings:
      local_mount: /home

### STARTUP SOFTWARE ###

  - id: startup
    source: modules/scripts/startup-script
    settings:
      runners:
          - type: shell
          destination: "decompress.sh"
          content: |
            #!/bin/sh
            echo $2
            tar zxvf /tmp/$1 -C /
          args: "bar.tgz 'Expanding file'"


### Slurm Partitions and Nodes ###

  - id: debug_node_group
    source: community/modules/compute/schedmd-slurm-gcp-v5-node-group
    settings:
      node_count_dynamic_max: 4
      machine_type: n2-standard-2
      instance_image_custom: true
      instance_image:
        family: my-custom-slurm
        project: psivant-dev

  - id: debug_partition
    source: community/modules/compute/schedmd-slurm-gcp-v5-partition
    use:
    - network1
    - homefs
    - debug_node_group
    settings:
      partition_name: debug
      exclusive: false # allows nodes to stay up after jobs are done
      enable_placement: false # the default is: true
      is_default: true

  - id: compute_node_group
    source: community/modules/compute/schedmd-slurm-gcp-v5-node-group
    settings:
      node_count_dynamic_max: 20
      bandwidth_tier: gvnic_enabled
      instance_image_custom: true
      instance_image:
        family: my-custom-slurm
        project: psivant-dev

  - id: compute_partition
    source: community/modules/compute/schedmd-slurm-gcp-v5-partition
    use:
    - network1
    - homefs
    - compute_node_group
    settings:
      partition_name: compute

  - id: gpu_node_group
    source: community/modules/compute/schedmd-slurm-gcp-v5-node-group
    settings:
      enable_smt: true
      node_count_dynamic_max: 20
      machine_type: g2-standard-24
      instance_image_custom: true
      instance_image:
        family: my-custom-slurm
        project: psivant-dev

  - id: gpu_partition
    source: community/modules/compute/schedmd-slurm-gcp-v5-partition
    use:
    - network1
    - gpu_node_group
    - homefs
    settings:
      partition_name: gpu


  - id: slurm_controller
    source: community/modules/scheduler/schedmd-slurm-gcp-v5-controller
    use:
    - network1
    - debug_partition
    - compute_partition
    - gpu_partition
    - homefs
    settings:
      disable_controller_public_ips: false
      instance_image_custom: true
      instance_image:
        family: my-custom-slurm
        project: psivant-dev

  - id: slurm_login
    source: community/modules/scheduler/schedmd-slurm-gcp-v5-login
    use:
    - network1
    - slurm_controller
    # - startup
    settings:
      machine_type: n2-standard-4
      disable_login_public_ips: false
      instance_image_custom: true
      instance_image:
        family: my-custom-slurm
        project: psivant-dev

```

**Comment by tomschultz-roivant** [comment]
Ok I was able to get pmix working by adding the `--with-pmix=/usr/local/pmix` to my fork of `slurm-gcp` and then pulling that in the build script you shared.

see: https://github.com/tomschultz-roivant/slurm-gcp/blob/d87dfbde4e6f25b5a7d25b39c1328649f6db4b35/ansible/roles/slurm/tasks/install.yml#L54

![image](https://github.com/GoogleCloudPlatform/hpc-toolkit/assets/87714410/0cadeed7-e440-4e3a-a984-0a42be636ba0)

**Comment by tpdownes** [comment]
I'll review why /usr/local appears to be necessary as the RPM description doesn't suggest it's putting any files under that path. Slurm 23.02 and below should autodiscover PMIx. It may be necessary or helpful to install pmix-devel package for autodiscovery. I am glad your are working for now, but we plan to create a reliable solution for both 23.02 and 23.11 (where autodiscovery has been disabled by SchedMD).

**Comment by tomschultz-roivant** [comment]
Sound good!  If there's anything I can do to help just let me know, I'm happy to test things or work on some tasks.  It's been a little bit since I contributed to OSS so I'm due to give back 😄

**Comment by tpdownes** [comment]
You might look at https://github.com/GoogleCloudPlatform/slurm-gcp/pull/53 for our official implementation in the Slurm-GCP v5 modules. We haven't yet published a public image based on this change, but you can use this implementation with:

```
          ansible-pull \
              -U https://github.com/GoogleCloudPlatform/slurm-gcp -C v5 \
              -i localhost, --limit localhost --connection=local \
              -e @/var/tmp/slurm_vars.json \
              ansible/playbook.yml
```

**Comment by tpdownes** [comment]
@tomschultz-roivant... we are expecting this support to be in the standard Slurm images in their next release (likely mid-April). A standard build on the latest tag will include this (in the preceding comment it was just on the v5 branch, but a git tag is part of the release process)

```
          ansible-pull \
              -U https://github.com/GoogleCloudPlatform/slurm-gcp -C 5.10.7 \
              -i localhost, --limit localhost --connection=local \
              -e @/var/tmp/slurm_vars.json \
              ansible/playbook.yml
```

**Comment by tomschultz-roivant** [comment]
Thanks! I will try this out today!

**Comment by tomschultz-roivant** [comment]
This worked great! I only tried it with Centos 7 so far. I'm also trying it with Ubuntu.

**Comment by tpdownes** [comment]
Default Slurm images with PMIx are now available on our develop branch in the Slurm-GCP 6.5.2 (#2594) and Slurm-GCP 5.11.1 releases (#2595)
